File: frontend/lib/src/core/api/models/research.dart
// Created with the Assistance of Claude Code
// frontend/lib/src/core/api/models/research.dart
/// Research data models matching backend Pydantic schemas
///
/// Run `dart run build_runner build` to generate .g.dart and .freezed.dart files
library;

import 'package:freezed_annotation/freezed_annotation.dart';

part 'research.freezed.dart';
part 'research.g.dart';

/// Survey item in the research survey list
@freezed
sealed class ResearchSurvey with _$ResearchSurvey {
  const factory ResearchSurvey({
    @JsonKey(name: 'survey_id') required int surveyId,
    required String title,
    @JsonKey(name: 'publication_status') required String publicationStatus,
    @JsonKey(name: 'response_count') required int responseCount,
    @JsonKey(name: 'question_count') required int questionCount,
  }) = _ResearchSurvey;

  factory ResearchSurvey.fromJson(Map<String, dynamic> json) =>
      _$ResearchSurveyFromJson(json);
}

/// Survey overview stats returned by GET /research/surveys/{id}/overview
@freezed
sealed class SurveyOverview with _$SurveyOverview {
  const factory SurveyOverview({
    @JsonKey(name: 'survey_id') required int surveyId,
    required String title,
    @JsonKey(name: 'respondent_count') required int respondentCount,
    @JsonKey(name: 'completion_rate') required double completionRate,
    @JsonKey(name: 'question_count') required int questionCount,
    required bool suppressed,
    @JsonKey(name: 'min_responses') @Default(5) int minResponses,
  }) = _SurveyOverview;

  factory SurveyOverview.fromJson(Map<String, dynamic> json) =>
      _$SurveyOverviewFromJson(json);
}

/// Aggregate data for a single question
@freezed
sealed class QuestionAggregate with _$QuestionAggregate {
  const factory QuestionAggregate({
    @JsonKey(name: 'question_id') required int questionId,
    @JsonKey(name: 'question_content') required String questionContent,
    @JsonKey(name: 'response_type') required String responseType,
    String? category,
    @JsonKey(name: 'response_count') required int responseCount,
    required bool suppressed,
    Map<String, dynamic>? data,
  }) = _QuestionAggregate;

  factory QuestionAggregate.fromJson(Map<String, dynamic> json) =>
      _$QuestionAggregateFromJson(json);
}

/// Full aggregate response for a survey (list of question aggregates)
@freezed
sealed class AggregateResponse with _$AggregateResponse {
  const factory AggregateResponse({
    @JsonKey(name: 'survey_id') required int surveyId,
    required String title,
    @JsonKey(name: 'total_respondents') required int totalRespondents,
    required List<QuestionAggregate> aggregates,
  }) = _AggregateResponse;

  factory AggregateResponse.fromJson(Map<String, dynamic> json) =>
      _$AggregateResponseFromJson(json);
}

/// Question metadata in individual response data
@freezed
sealed class ResponseQuestion with _$ResponseQuestion {
  const factory ResponseQuestion({
    @JsonKey(name: 'question_id') required int questionId,
    @JsonKey(name: 'question_content') required String questionContent,
    @JsonKey(name: 'response_type') required String responseType,
    String? category,
  }) = _ResponseQuestion;

  factory ResponseQuestion.fromJson(Map<String, dynamic> json) =>
      _$ResponseQuestionFromJson(json);
}

/// A single anonymized participant's response row
@freezed
sealed class ResponseRow with _$ResponseRow {
  const factory ResponseRow({
    @JsonKey(name: 'anonymous_id') required String anonymousId,
    required Map<String, String> responses,
  }) = _ResponseRow;

  factory ResponseRow.fromJson(Map<String, dynamic> json) =>
      _$ResponseRowFromJson(json);
}

/// Individual anonymized response data for a survey
@freezed
sealed class IndividualResponseData with _$IndividualResponseData {
  const factory IndividualResponseData({
    @JsonKey(name: 'survey_id') required int surveyId,
    required String title,
    @JsonKey(name: 'respondent_count') required int respondentCount,
    required bool suppressed,
    String? reason,
    required List<ResponseQuestion> questions,
    required List<ResponseRow> rows,
  }) = _IndividualResponseData;

  factory IndividualResponseData.fromJson(Map<String, dynamic> json) =>
      _$IndividualResponseDataFromJson(json);
}

// ---------------------------------------------------------------------------
// Cross-survey models
// ---------------------------------------------------------------------------

/// Question metadata in cross-survey response data (includes survey info)
@freezed
sealed class CrossSurveyQuestion with _$CrossSurveyQuestion {
  const factory CrossSurveyQuestion({
    @JsonKey(name: 'question_id') required int questionId,
    @JsonKey(name: 'question_content') required String questionContent,
    @JsonKey(name: 'response_type') required String responseType,
    String? category,
    @JsonKey(name: 'survey_id') required int surveyId,
    @JsonKey(name: 'survey_title') required String surveyTitle,
    @JsonKey(name: 'survey_start_date') String? surveyStartDate,
  }) = _CrossSurveyQuestion;

  factory CrossSurveyQuestion.fromJson(Map<String, dynamic> json) =>
      _$CrossSurveyQuestionFromJson(json);
}

/// A single anonymized participant's response row (cross-survey)
/// One row per participant with all responses merged across surveys.
@freezed
sealed class CrossSurveyRow with _$CrossSurveyRow {
  const factory CrossSurveyRow({
    @JsonKey(name: 'anonymous_id') required String anonymousId,
    required Map<String, String> responses,
  }) = _CrossSurveyRow;

  factory CrossSurveyRow.fromJson(Map<String, dynamic> json) =>
      _$CrossSurveyRowFromJson(json);
}

/// Per-survey summary in cross-survey response
@freezed
sealed class CrossSurveySummary with _$CrossSurveySummary {
  const factory CrossSurveySummary({
    @JsonKey(name: 'survey_id') required int surveyId,
    required String title,
    @JsonKey(name: 'respondent_count') required int respondentCount,
  }) = _CrossSurveySummary;

  factory CrossSurveySummary.fromJson(Map<String, dynamic> json) =>
      _$CrossSurveySummaryFromJson(json);
}

/// Cross-survey overview stats
@freezed
sealed class CrossSurveyOverview with _$CrossSurveyOverview {
  const factory CrossSurveyOverview({
    @JsonKey(name: 'survey_ids') required List<int> surveyIds,
    required List<CrossSurveySummary> surveys,
    @JsonKey(name: 'total_respondent_count') required int totalRespondentCount,
    @JsonKey(name: 'total_question_count') required int totalQuestionCount,
    @JsonKey(name: 'avg_completion_rate') required double avgCompletionRate,
    required bool suppressed,
    String? reason,
    @JsonKey(name: 'min_responses') @Default(5) int minResponses,
  }) = _CrossSurveyOverview;

  factory CrossSurveyOverview.fromJson(Map<String, dynamic> json) =>
      _$CrossSurveyOverviewFromJson(json);
}

/// Cross-survey aggregate stats (Data Bank analysis tab)
@freezed
sealed class CrossSurveyAggregateResponse with _$CrossSurveyAggregateResponse {
  const factory CrossSurveyAggregateResponse({
    @JsonKey(name: 'survey_ids') required List<int> surveyIds,
    @JsonKey(name: 'total_respondents') required int totalRespondents,
    required List<QuestionAggregate> aggregates,
  }) = _CrossSurveyAggregateResponse;

  factory CrossSurveyAggregateResponse.fromJson(Map<String, dynamic> json) =>
      _$CrossSurveyAggregateResponseFromJson(json);
}

/// Cross-survey individual response data
@freezed
sealed class CrossSurveyResponseData with _$CrossSurveyResponseData {
  const factory CrossSurveyResponseData({
    @JsonKey(name: 'survey_ids') required List<int> surveyIds,
    required List<CrossSurveySummary> surveys,
    @JsonKey(name: 'total_respondent_count') required int totalRespondentCount,
    @JsonKey(name: 'date_from') String? dateFrom,
    @JsonKey(name: 'date_to') String? dateTo,
    required bool suppressed,
    String? reason,
    @JsonKey(name: 'suppressed_surveys') @Default([]) List<int> suppressedSurveys,
    required List<CrossSurveyQuestion> questions,
    required List<CrossSurveyRow> rows,
  }) = _CrossSurveyResponseData;

  factory CrossSurveyResponseData.fromJson(Map<String, dynamic> json) =>
      _$CrossSurveyResponseDataFromJson(json);
}

File: backend/app/api/v1/research.py
# Created with the Assistance of Claude Code
# backend/app/api/v1/research.py
"""
Research Data API — Individual Anonymized Responses & Aggregate Analytics

Endpoints:
- GET  /surveys                           - List surveys with response counts
- GET  /surveys/{id}/overview             - Survey overview stats
- GET  /surveys/{id}/responses            - Individual anonymized response rows
- GET  /surveys/{id}/aggregates           - All question aggregates (filterable)
- GET  /surveys/{id}/aggregates/{qid}     - Single question aggregate
- GET  /surveys/{id}/export/csv           - CSV download of individual responses
- GET  /cross-survey/overview             - Data bank overview stats
- GET  /cross-survey/questions            - Available questions for field picker
- GET  /cross-survey/responses            - Data bank individual responses
- GET  /cross-survey/aggregates           - Data bank aggregate stats for charts
- GET  /cross-survey/export/csv           - Data bank CSV download

All endpoints require researcher (RoleID=2) or admin (RoleID=4) role.
Individual responses use anonymized participant IDs (SHA-256 hash).
Data is suppressed if fewer than 5 distinct respondents (k-anonymity).

SECURITY: All queries use parameterized statements to prevent SQL injection.
"""

from fastapi import APIRouter, Depends, HTTPException, Query
from fastapi.responses import StreamingResponse
from pydantic import BaseModel
from typing import Optional
import io

from ...utils.utils import get_db_connection
from ..deps import require_role
from ...services.aggregation import AggregationService
from ...services.settings import get_int_setting

router = APIRouter()
_aggregation = AggregationService()


# ---------------------------------------------------------------------------
# Pydantic response models
# ---------------------------------------------------------------------------

class ResearchSurvey(BaseModel):
    survey_id: int
    title: str
    publication_status: str
    response_count: int
    question_count: int


class SurveyOverviewResponse(BaseModel):
    survey_id: int
    title: str
    respondent_count: int
    completion_rate: float
    question_count: int
    suppressed: bool
    reason: Optional[str] = None
    min_responses: int = 5


class QuestionAggregateResponse(BaseModel):
    question_id: int
    question_content: str
    response_type: str
    category: Optional[str] = None
    response_count: int
    suppressed: bool
    reason: Optional[str] = None
    data: Optional[dict] = None


class AggregateListResponse(BaseModel):
    survey_id: int
    title: str
    total_respondents: int
    aggregates: list[QuestionAggregateResponse]


class ResponseQuestion(BaseModel):
    question_id: int
    question_content: str
    response_type: str
    category: Optional[str] = None


class ResponseRow(BaseModel):
    anonymous_id: str
    responses: dict[str, str]


class IndividualResponseData(BaseModel):
    survey_id: int
    title: str
    respondent_count: int
    suppressed: bool
    reason: Optional[str] = None
    questions: list[ResponseQuestion]
    rows: list[ResponseRow]


# Cross-survey models

class CrossSurveyQuestion(BaseModel):
    question_id: int
    question_content: str
    response_type: str
    category: Optional[str] = None
    survey_id: int
    survey_title: str
    survey_start_date: Optional[str] = None


class CrossSurveyRow(BaseModel):
    anonymous_id: str
    responses: dict[str, str]


class CrossSurveySummary(BaseModel):
    survey_id: int
    title: str
    respondent_count: int


class CrossSurveyOverviewResponse(BaseModel):
    survey_ids: list[int]
    surveys: list[CrossSurveySummary]
    total_respondent_count: int
    total_question_count: int
    avg_completion_rate: float
    suppressed: bool
    reason: Optional[str] = None
    min_responses: int = 5


class CrossSurveyResponseData(BaseModel):
    survey_ids: list[int]
    surveys: list[CrossSurveySummary]
    total_respondent_count: int
    date_from: Optional[str] = None
    date_to: Optional[str] = None
    suppressed: bool
    reason: Optional[str] = None
    suppressed_surveys: list[int] = []
    questions: list[CrossSurveyQuestion]
    rows: list[CrossSurveyRow]


class CrossSurveyAggregateResponse(BaseModel):
    survey_ids: list[int]
    total_respondents: int
    aggregates: list[QuestionAggregateResponse]


# ---------------------------------------------------------------------------
# Endpoints
# ---------------------------------------------------------------------------

@router.get("/surveys", response_model=list[ResearchSurvey])
def list_research_surveys(user=Depends(require_role(2, 4))):
    """List all surveys with response counts for the researcher."""
    conn = get_db_connection()
    cur = conn.cursor(dictionary=True)
    try:
        cur.execute(
            """
            SELECT
                s.SurveyID,
                s.Title,
                s.PublicationStatus,
                COUNT(DISTINCT r.ParticipantID) AS response_count,
                (SELECT COUNT(*) FROM QuestionList ql WHERE ql.SurveyID = s.SurveyID) AS question_count
            FROM Survey s
            LEFT JOIN Responses r ON r.SurveyID = s.SurveyID
            GROUP BY s.SurveyID, s.Title, s.PublicationStatus
            ORDER BY s.SurveyID DESC
            """
        )
        rows = cur.fetchall()
        return [
            ResearchSurvey(
                survey_id=row["SurveyID"],
                title=row["Title"],
                publication_status=row["PublicationStatus"],
                response_count=row["response_count"],
                question_count=row["question_count"],
            )
            for row in rows
        ]
    finally:
        cur.close()
        conn.close()


@router.get("/surveys/{survey_id}/overview", response_model=SurveyOverviewResponse)
def get_survey_overview(survey_id: int, user=Depends(require_role(2, 4))):
    """Get high-level stats for a survey."""
    result = _aggregation.get_survey_overview(survey_id)
    if result is None:
        raise HTTPException(status_code=404, detail="Survey not found")
    return SurveyOverviewResponse(**result, min_responses=get_int_setting("k_anonymity_threshold"))


@router.get("/surveys/{survey_id}/responses", response_model=IndividualResponseData)
def get_individual_responses(
    survey_id: int,
    category: Optional[str] = Query(None, description="Filter by question category"),
    response_type: Optional[str] = Query(None, description="Filter by response type"),
    user=Depends(require_role(2, 4)),
):
    """Get individual anonymized response rows for a survey."""
    result = _aggregation.get_individual_responses(
        survey_id, category=category, response_type=response_type
    )
    if result is None:
        raise HTTPException(status_code=404, detail="Survey not found")
    return IndividualResponseData(**result)


@router.get("/surveys/{survey_id}/aggregates", response_model=AggregateListResponse)
def get_survey_aggregates(
    survey_id: int,
    category: Optional[str] = Query(None, description="Filter by question category"),
    response_type: Optional[str] = Query(None, description="Filter by response type"),
    user=Depends(require_role(2, 4)),
):
    """Get aggregate data for all questions in a survey."""
    result = _aggregation.get_survey_aggregates(
        survey_id, category=category, response_type=response_type
    )
    if result is None:
        raise HTTPException(status_code=404, detail="Survey not found")
    return AggregateListResponse(**result)


@router.get(
    "/surveys/{survey_id}/aggregates/{question_id}",
    response_model=QuestionAggregateResponse,
)
def get_question_aggregate(
    survey_id: int, question_id: int, user=Depends(require_role(2, 4))
):
    """Get aggregate data for a single question."""
    result = _aggregation.get_question_aggregate(survey_id, question_id)
    if result is None:
        raise HTTPException(status_code=404, detail="Question not found in survey")
    return QuestionAggregateResponse(**result)


@router.get("/surveys/{survey_id}/export/csv")
def export_csv(
    survey_id: int,
    category: Optional[str] = Query(None),
    response_type: Optional[str] = Query(None),
    user=Depends(require_role(2, 4)),
):
    """Export individual anonymized response data as CSV download."""
    # Look up survey title for CSV metadata
    overview = _aggregation.get_survey_overview(survey_id)
    title = overview["title"] if overview else None
    csv_data = _aggregation.get_individual_csv_export(
        survey_id, category=category, response_type=response_type,
        survey_title=title,
    )
    if csv_data is None:
        raise HTTPException(status_code=404, detail="Survey not found")

    # Use survey title in filename (sanitized)
    safe_title = (title or f"survey_{survey_id}").replace(" ", "_").replace("/", "_")
    return StreamingResponse(
        io.StringIO(csv_data),
        media_type="text/csv",
        headers={
            "Content-Disposition": f"attachment; filename={safe_title}_responses.csv"
        },
    )


# ---------------------------------------------------------------------------
# Cross-survey endpoints
# ---------------------------------------------------------------------------

@router.get("/cross-survey/overview", response_model=CrossSurveyOverviewResponse)
def get_cross_survey_overview(
    survey_ids: Optional[list[int]] = Query(None, description="Survey IDs (omit for all)"),
    date_from: Optional[str] = Query(None, description="Start date YYYY-MM-DD"),
    date_to: Optional[str] = Query(None, description="End date YYYY-MM-DD"),
    user=Depends(require_role(2, 4)),
):
    """Get overview stats across surveys. Omit survey_ids for all surveys."""
    result = _aggregation.get_cross_survey_overview(
        survey_ids or None, date_from=date_from, date_to=date_to
    )
    return CrossSurveyOverviewResponse(**result, min_responses=get_int_setting("k_anonymity_threshold"))


@router.get("/cross-survey/questions", response_model=list[CrossSurveyQuestion])
def get_available_questions(
    survey_ids: Optional[list[int]] = Query(None, description="Filter by survey IDs"),
    category: Optional[str] = Query(None, description="Filter by question category"),
    response_type: Optional[str] = Query(None, description="Filter by response type"),
    user=Depends(require_role(2, 4)),
):
    """List all available questions in the data bank for field selection."""
    result = _aggregation.get_available_questions(
        survey_ids or None, category=category, response_type=response_type
    )
    return [CrossSurveyQuestion(**q) for q in result]


@router.get("/cross-survey/responses", response_model=CrossSurveyResponseData)
def get_cross_survey_responses(
    survey_ids: Optional[list[int]] = Query(None, description="Survey IDs (omit for all)"),
    question_ids: Optional[list[int]] = Query(None, description="Specific question IDs"),
    category: Optional[str] = Query(None, description="Filter by question category"),
    response_type: Optional[str] = Query(None, description="Filter by response type"),
    date_from: Optional[str] = Query(None, description="Start date YYYY-MM-DD"),
    date_to: Optional[str] = Query(None, description="End date YYYY-MM-DD"),
    user=Depends(require_role(2, 4)),
):
    """Get individual anonymized response rows. Omit survey_ids for all surveys."""
    result = _aggregation.get_cross_survey_responses(
        survey_ids or None,
        date_from=date_from,
        date_to=date_to,
        category=category,
        response_type=response_type,
        question_ids=question_ids or None,
    )
    return CrossSurveyResponseData(**result)


@router.get("/cross-survey/aggregates", response_model=CrossSurveyAggregateResponse)
def get_cross_survey_aggregates(
    survey_ids: Optional[list[int]] = Query(None, description="Survey IDs (omit for all)"),
    question_ids: Optional[list[int]] = Query(None, description="Specific question IDs"),
    category: Optional[str] = Query(None, description="Filter by question category"),
    response_type: Optional[str] = Query(None, description="Filter by response type"),
    date_from: Optional[str] = Query(None, description="Start date YYYY-MM-DD"),
    date_to: Optional[str] = Query(None, description="End date YYYY-MM-DD"),
    user=Depends(require_role(2, 4)),
):
    """Get aggregate stats across the data bank for charting."""
    result = _aggregation.get_cross_survey_aggregates(
        survey_ids or None,
        question_ids=question_ids or None,
        date_from=date_from,
        date_to=date_to,
        category=category,
        response_type=response_type,
    )
    return CrossSurveyAggregateResponse(**result)


@router.get("/cross-survey/export/csv")
def export_cross_survey_csv(
    survey_ids: Optional[list[int]] = Query(None, description="Survey IDs (omit for all)"),
    question_ids: Optional[list[int]] = Query(None, description="Specific question IDs"),
    category: Optional[str] = Query(None),
    response_type: Optional[str] = Query(None),
    date_from: Optional[str] = Query(None),
    date_to: Optional[str] = Query(None),
    user=Depends(require_role(2, 4)),
):
    """Export data bank individual response data as CSV download."""
    csv_data = _aggregation.get_cross_survey_csv_export(
        survey_ids or None,
        date_from=date_from,
        date_to=date_to,
        category=category,
        response_type=response_type,
        question_ids=question_ids or None,
    )
    if csv_data is None:
        raise HTTPException(status_code=404, detail="No surveys found")

    return StreamingResponse(
        io.StringIO(csv_data),
        media_type="text/csv",
        headers={
            "Content-Disposition": "attachment; filename=data_bank_export.csv"
        },
    )
